server/src/linter.ts

```diff
@@ -0,0 +1,199 @@
+import { spawn } from 'child_process'
+import * as LSP from 'vscode-languageserver'
+
+function formatMessage(comment: ShellcheckComment): string {
+  return (comment.code ? `SC${comment.code}: ` : '') + comment.message
+}
+
+type LinterOptions = {
+  executablePath?: string | null
+  cwd?: string
+}
+
+export default class Linter {
+  private executablePath: string | null
+  private cwd: string
+  _canLint: boolean
+
+  constructor(opts: LinterOptions = {}) {
+    this.executablePath = opts.executablePath || null
+    this.cwd = opts.cwd || process.cwd()
+    this._canLint = !!this.executablePath
+  }
+
+  public get canLint(): boolean {
+    return this._canLint
+  }
+
+  public async lint(
+    document: LSP.TextDocument,
+    folders: LSP.WorkspaceFolder[],
+  ): Promise<LSP.Diagnostic[]> {
+    if (!this.executablePath || !this._canLint) return []
+
+    const result = await this.runShellcheck(this.executablePath, document, folders)
+    if (!this._canLint) return []
+
+    const diags: LSP.Diagnostic[] = []
+    for (const comment of result.comments) {
+      const start: LSP.Position = {
+        line: comment.line - 1,
+        character: comment.column - 1,
+      }
+      const end: LSP.Position = {
+        line: comment.endLine - 1,
+        character: comment.endColumn - 1,
+      }
+
+      diags.push({
+        message: formatMessage(comment),
+        severity: mapShellcheckServerity(comment.level),
+        code: comment.code,
+        source: 'shellcheck',
+        range: { start, end },
+      })
+    }
+
+    return diags
+  }
+
+  private async runShellcheck(
+    executablePath: string,
+    document: LSP.TextDocument,
+    folders: LSP.WorkspaceFolder[],
+  ): Promise<ShellcheckResult> {
+    // FIXME: inject?
```

Comment:
What does this FIXME mean?

Reply:
It looks like a remnant. I think it was there before I got workspace folders connected. I'll remove it now.